lib/plugins/blocks.js

```diff
@@ -81,6 +81,32 @@ function inject (bot, { version, storageBuilder }) {
     bot.world.unloadColumn(chunkX, chunkZ)
   }
 
+  // Allocates new chunk object while taking world's custom min/max height into account
+  function createNewChunk () {
+    const chunkInitializationData = {}
+    const dimensionData = bot.game.dimensionData
+
+    if (dimensionData) {
+      chunkInitializationData.minWorldHeight = dimensionData.min_y
+      chunkInitializationData.maxWorldHeight = dimensionData.min_y + dimensionData.height
+    }
+
+    return new Chunk(chunkInitializationData)
+  }
+
+  function flattenMaskArray (sectionBitMask) {
+    // We need to handle arrays specially because protocol will return
+    // array of BigInt on new versions, while we really need an array of 32-bit integers,
```

Comment:
where does the protocol return BigInts?

Reply:
In the primaryBitMask field of packet_map_chunk of 1.17 protocol. All chunk related mask fields are now serialized as long arrays representing bit sets in fact, including lighting packet ones. That's needed to support extended height.

Reply:
I think the format of chunk packet in protocol 1.17.json is wrong, we should change it so things like this are not required.

Reply:
I looked through the Minecraft java source and it is writing it as a 64-bit big endian array, so I don't get what this flipping this is about.
```java
         var1.writeLongArray(this.storage.getRaw()); // PalettedContainer.java
```
```java
   public FriendlyByteBuf writeLongArray(long[] var1) { // FriendlyByteBuf.java
      this.writeVarInt(var1.length);
      long[] var2 = var1;
      int var3 = var1.length;

      for(int var4 = 0; var4 < var3; ++var4) {
         long var5 = var2[var4];
         this.writeLong(var5);
      }

      return this;
   }
   ...
   public ByteBuf writeLong(long var1) {
      return this.source.writeLong(var1);
   }

   public ByteBuf writeLongLE(long var1) {
      return this.source.writeLongLE(var1);
   }
```

Reply:
this should at worse be happening in pchunk, not here
at best be unneeded by improving the definition of the packet in protocol.json

Reply:
Flipping is needed because BitSet implementation in prismarine-chunk requires least significant bits to come first pretty much because bit addressing starts from zero and grows upwards.
Original implementation in Java doesn't need to do this because it actually operates on 64-bit integers and not 32-bit ones, but in JS we're kinda forced to use 32-bit numbers.

Reply:
solved this